core/domain/exp_domain.py

```diff
@@ -1298,6 +1316,26 @@ def is_demo(self):
         """
         return self.is_demo_exploration_id(self.id)
 
+    @property
+    def is_stale(self):
```

Comment:
This is wrong, it should be

```
@property
def proto_size_in_bytes(self):
    ...
```

and self._is_stale should be a private member set in `__init__`.

That way you can refer to self.proto_size_in_bytes and always know that you are getting the right value when doing so.

Reply:
As you had mentioned in the comment below `You do not have to update self.android_proto_size_in_bytes here. That only gets updated when the value is read (see above).` 
What I am understanding from this is android_proto_size_in_bytes gets set in init function only like
```py
self.android_proto_size_in_bytes = self.proto_size_in_bytes
```

What does `self._is_stale` do?

Reply:
No, that's not what my comment means. My comment means that we recalculate self.android_proto_size_in_bytes when something accesses self.android_proto_size_in_bytes (which will call the method decorated with `@property`, which will return a value based on is_stale and the cached proto size). You might want to read up more about `@property` if you're not familiar with it.

So, in this class, self.android_proto_size_in_bytes is kind of a pseudo-property -- it's really a method behind the scenes and its calculation always involves some logic. You do not need to ever set it.

is_stale is used in the calculation of self.android_proto_size_in_bytes. It starts off as True. In the calculation of self.android_proto_size_in_bytes, it is used to determine whether or not the cached proto size needs to be updated before the value is returned to the caller.

Reply:
If it's helpful, an alternative is to think about self.android_proto_size_in_bytes as self.get_android_proto_size_in_bytes() instead and don't assume the existence of any self.android_proto_size_in_bytes class member. That's really what is going on here at a conceptual level.

Reply:
Ah, The `self._is_stale` mentioned is our `self.android_proto_size_is_stale = True`. 
I thought you were expecting an another new member.

Reply:
Yes, that's correct, they're the same thing. Sorry for the confusion.

Reply:
> self.android_proto_size_in_bytes as self.get_android_proto_size_in_bytes()

1. As you had mentioned keeping this field as read-only and the calculation happens only at the access time, then how do we set this field in the `to_dict` and `form_dict`? Is the below-mentioned approach is correct?

```py
exploration._android_proto_size_in_bytes = (
            exploration_dict['android_proto_size_in_bytes'])

exploration_dict['android_proto_size_in_bytes'] = (
            exploration.android_proto_size_in_bytes)

'android_proto_size_in_bytes': self.android_proto_size_in_bytes,
```

2. In the `_init_` function, we can either ignore initializing `self._android_proto_size_in_bytes` or initialize with:
```py
self._android_proto_size_in_bytes = self.android_proto_size_in_bytes
```

Reply:
1. You would not need to set `_android_proto_size_in_bytes` in the exploration object. For from_dict(), I think we would just ignore the proto_size_in_bytes value that exists in the dict, since we'd calculate it based on the rest. We might want to throw/log a non-blocking error if the dict proto size does not match the computed proto size though, since that indicates that something might be wrong.

For to_dict(), I have a question -- is the resulting dict just used in the Web frontend? If so, will the Web frontend actually ever need to know the proto size in bytes, or is this info only for the proto export? If the latter, then you don't actually need to include this field in from_dict and to_dict IMO. That would also probably make things simpler because we don't have to keep that value updated in the Web frontend.

2. As mentioned before, do not ever set self._android_proto_size_in_bytes or self.android_proto_size_in_bytes, in initialization or otherwise. Only the cached size is stored; otherwise, every reference to android_proto_size_in_bytes should actually be doing a calculation using the `@property` method (which you can think of as a "get" method). Initially the stale flag is set to True and the cached size is 0, forcing a recomputation the next time that property is accessed.

Reply:
Fixed

Reply:
@seanlip could you take an another pass on this and confirm if the proto size updation is correct now?